hasktorch/src/Torch/Typed/Autograd.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+module Torch.Typed.Autograd where
+
+import           Data.HList
+import           GHC.TypeLits
+import           System.IO.Unsafe
+
+import qualified ATen.Cast as ATen
+import qualified ATen.Class as ATen
+import qualified Torch.Managed.Autograd
+import qualified Torch.DType                   as D
+import qualified Torch.Device                  as D
+import qualified Torch.Tensor                  as D
+import           Torch.Typed.Tensor
+import           Torch.Typed.Parameter
+
+type family GradR (parameters :: [a]) (dtype :: D.DType) (device :: (D.DeviceType, Nat)) :: [a] where
```

Comment:
Why do we require all gradients to be on a single device and have the same dtype? This seems like an unnecessary requirement

Reply:
Hey @apaszke, thanks for the question. This is interesting, are you implying that `grad` supports dtype promotion and loss and/or tensor parameters on different devices? I never tried it on such inputs. I just assumed it would be as restricted as most other ops.

Reply:
Not sure what you mean by "dtype promotion", but the quantities you differentiate _can_ be on different devices and have distinct dtypes. The returned gradients will have the dtypes and devices matching the HList of quantities differentiated wrt., so they can be heterogenous too.

Reply:
dtype promotion as described here: https://pytorch.org/docs/stable/tensor_attributes.html

This is great news for me, that means that a severe limitation I had taken for granted actually doesn’t exist.

Reply:
No, data types are always preserved. The signature of grad really is:
```
grad :: outputTensors -> inputTensors -> inputTensors
```
meaning that the gradients are a HList of the same type as the second argument.

Reply:
Nice, I’ll lift the restrictions and add some tests in my next PR